beacon-light-client/plonky2/input_fetchers/balance_verification/deposits_accumulator/lib/scheduler.ts

```diff
@@ -0,0 +1,161 @@
+    this.provider = new ethers.providers.JsonRpcProvider(options['rpc-url']);
+    this.contract = new ethers.Contract(
+      options['address'],
+      ValidatorsAccumulator.abi,
+      this.provider,
+    );
+
+    let latestLoggedBlock = await this.redis.get(constants.latestBlock);
+    if (!latestLoggedBlock) {
+      latestLoggedBlock = (
+        (await this.provider.getBlockNumber()) - EVENTS_BATCH_SIZE
+      ).toString();
+    }
+    this.syncBlock = options['sync-block'] || +latestLoggedBlock;
+
+    const mod = await import('@lodestar/types');
+    this.ssz = mod.ssz;
+  }
+
+  async dispose() {
+    return this.redis.quit();
+  }
+
+  async start() {
+    // Sync to latest block
+    await this.syncToLatestBlock();
+
+    console.log(chalk.bold.blue('Listening for events...'));
+    // Start listening to head events
+    this.contract.on(Events.Deposited, async (...args: any) => {
+      const data = args[args.length - 1];
+
+      if (this.syncBlock <= data.blockNumber) {
+        await this.updateDepositsBatch([
+          {
+            event: {
+              pubkey: data.args.pubkey,
+              depositIndex: data.args.depositIndex,
+              signature: data.args.signature,
+              depositMessageRoot: data.args.depositMessageRoot,
+            },
+            blockNumber: data.blockNumber,
+          },
+        ]);
+      }
+    });
+  }
+
+  async syncToLatestBlock() {
+    const latestBlock = await this.provider.getBlockNumber();
+    console.log(
+      chalk.bold.blue(`Syncing to block (${chalk.cyan(latestBlock)})...`),
+    );
+
+    if (this.syncBlock > latestBlock) {
+      return;
+    }
+
+    const iterations = Math.floor(
+      (latestBlock - this.syncBlock) / EVENTS_BATCH_SIZE,
+    );
+
+    let deposits: any[] = [];
+    for (let i = 0; i < iterations; i++) {
+      const logs = await getEvents(
+        this.provider,
+        this.contract,
+        {
+          [Events.Deposited]: [
+            'pubkey',
+            'depositIndex',
+            'signature',
+            'depositMessageRoot',
+          ],
+        },
+        this.syncBlock,
+        this.syncBlock + EVENTS_BATCH_SIZE - 1,
+      );
+
+      deposits.push(
+        ...logs.map(log => {
+          return {
+            event: log[Events.Deposited],
+            blockNumber: log[Events.Deposited].blockNumber,
+          };
+        }),
+      );
+
+      this.syncBlock += EVENTS_BATCH_SIZE;
+    }
+
+    const logs = await getEvents(
+      this.provider,
+      this.contract,
+      {
+        [Events.Deposited]: [
+          'pubkey',
+          'depositIndex',
+          'signature',
+          'depositMessageRoot',
+        ],
+      },
+      this.syncBlock,
+      latestBlock,
+    );
+
+    deposits.push(
+      ...logs.map(log => {
+        return {
+          event: log[Events.Deposited],
+          blockNumber: log[Events.Deposited].blockNumber,
+        };
+      }),
+    );
```

Comment:
See if this function does what you want
```ts
async function* fetchEvents(
  contract: ethers.Contract,
  event: string,
  firstBlock: number,
  lastBlock: number
): AsyncGenerator<ethers.Event[]> {
  const chunkSize = 10000;

  for (let block = firstBlock; block <= lastBlock; block += chunkSize) {
    const lastBlockInChunk = Math.min(block + chunkSize - 1, lastBlock);
    yield await contract.queryFilter(event, block, lastBlockInChunk);
  }
}
```

Reply:
`getEvents` already provides better event handling, but I could reuse the iteration logic  ✅